src/ipyradiant/basic_tools/object_literal_collapsing.py

```diff
@@ -0,0 +1,184 @@
+        """
+        # query for all predicates where object is not literal
+        predicates_to_uris = """
+            SELECT DISTINCT  ?p
+            WHERE {
+                    ?s ?p ?o .
+                FILTER (!isLiteral(?o))
+                }
+            """
+        q1 = self.graph.query(predicates_to_uris)
+        q1_results = set([result[0] for result in q1])
+
+        # query for predicates where object IS literal
+        predicates_to_literals = """
+            SELECT DISTINCT  ?p
+            WHERE {
+                    ?s ?p ?o .
+                FILTER isLiteral(?o)
+                }
+            """
+        q2 = self.graph.query(predicates_to_literals)
+
+        q2_results = set([result[0] for result in q2])
+        # take difference of sets and turn to list
+        selected_things = list(q2_results - q1_results)
+        available_things = list(q1_results - q2_results)
+        truncated_selected_uris = []
+        truncated_available_uris = []
+        for ref in selected_things:
+            truncated_selected_uris.append(
+                CustomURIRef(uri=ref, namespaces=self.graph.namespace_manager)
+            )
+        for ref in available_things:
+            truncated_available_uris.append(
+                CustomURIRef(uri=ref, namespaces=self.graph.namespace_manager)
+            )
+        # make sure things arent on both sides
+        for selected_item in self.multiselect.selected_things_list:
+            if selected_item not in truncated_selected_uris:
+                truncated_selected_uris.append(selected_item)
+                truncated_available_uris.remove(selected_item)
+
+        self.multiselect.available_things_list = truncated_available_uris
+        self.multiselect.selected_things_list = truncated_selected_uris
+
+
+def collapse_predicates(
+    graph: Graph,
+    predicates_to_collapse: list,
+    namespaces: Union[dict, NamespaceManager],
+) -> networkx.Graph:
+    """
+    This is a function to collapse the desired predicates of an RDF graph and return a collapsed version (networkx).
+
+    :params:
+    graph: an rdflib.graph.Graph Instance
+    predicates_to_collapse: a List of all the predicates (URIRefs) a user wants collapsed
+    namespaces: the namespaces to be used in the collapsed version, either as a Dict or an NamespaceManager object
+    """
+    # use external library functions
+    netx_graph = rdflib_to_networkx_digraph(graph)
+    # get list of subjects
+    subjects = [s for s, o in netx_graph.edges()]
+    objects_found = set()
+    for s, o in netx_graph.edges():
+
+        for _, p, _ in netx_graph[s][o]["triples"]:
+
+            if not isinstance(p, URIRef):
+                raise ValueError(f"Predicate must be URIRef not a '{type(p)}'.")
+            if p not in predicates_to_collapse:
+                continue
+
+            # add data to subject node
+            # TODO: custom representation?
```

Comment:
Can you expand on what you mean by this TODO (in the code file).

Reply:
actually thats already taken care of, forgot to get rid of it. my bad